File: process/Python/model/tax_calculator.py
from numpy import zeros_like, maximum, minimum
from pandas import DataFrame
from process.Python import TAX_BRACKETS


def cal_tax(income_series: DataFrame) -> DataFrame:

    total_tax = zeros_like(income_series, dtype=float)
    remaining_income = income_series.values.copy()

    for threshold, rate in TAX_BRACKETS:
        # Calculate income falling into this specific bracket
        taxable_at_rate = maximum(0, remaining_income - threshold)
        total_tax += taxable_at_rate * rate
        # Update remaining income to the threshold for the next bracket down
        remaining_income = minimum(remaining_income, threshold)

    return total_tax

File: process/R/model/tax_calculator.R
cal_tax <- function(income_series) {
  
  # Ensure input is numeric
  total_tax <- rep(0, length(income_series))
  remaining_income <- income_series
  
  # Assuming TAX_BRACKETS is a list of lists or a matrix/data frame 
  # with columns: threshold and rate
  # e.g., TAX_BRACKETS <- list(list(48000, 0.33), list(14000, 0.175), list(0, 0.105))
  for (bracket in TAX_BRACKETS) {
    threshold <- bracket[[1]]
    rate <- bracket[[2]]
    
    # Calculate income falling into this specific bracket
    # pmax is the element-wise version of max()
    taxable_at_rate <- pmax(0, remaining_income - threshold)
    total_tax <- total_tax + (taxable_at_rate * rate)
    
    # Update remaining income to the threshold for the next bracket down
    # pmin is the element-wise version of min()
    remaining_income <- pmin(remaining_income, threshold)
  }
  
  return(total_tax)
}